src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2240,227 @@ int64_t BlueFS::_read_random(
   return ret;
 }
 
+void BlueFS::_wal_update_size(FileRef file) {
+  uint64_t current_size = file->fnode.size;
+  uint64_t flush_offset = current_size;
+
+  FileReaderBuffer buf(super.block_size);
+
+  uint64_t x_off = 0;
+  uint64_t ret = 0;
+
+  std::unique_lock u_lock(file->lock);
+
+  while(flush_offset < file->fnode.allocated) {
+    // read first part of wal flush
+    {
+      buf.bl.clear();
+      buf.bl_off = p2align(flush_offset, (uint64_t)super.block_size);
+      auto p = file->fnode.seek(buf.bl_off, &x_off);
+      if (p == file->fnode.extents.end()) {
+        dout(5) << __func__ << " reading less then required "
+          << ret << "<" << ret << dendl;
+        break;
+      }
+      int r;
+      // TODO(pere): read extent length or max read size conf
+      if (!cct->_conf->bluefs_check_for_zeros) {
+        r = _bdev_read(p->bdev, p->offset + x_off, p->offset + p->length - x_off, &buf.bl, ioc[p->bdev],
+            false);
+      } else {
+        r = _read_and_check(
+            p->bdev, p->offset + x_off, p->offset + p->length - x_off, &buf.bl, ioc[p->bdev],
+            false);
+      }
+      ceph_assert(r == 0);
+      uint64_t flush_length = 0;
+      uint64_t data_left_on_buffer = buf.get_buf_remaining(flush_offset);
+      ceph_assert(data_left_on_buffer >= sizeof(File::WALFlush::WALLength));
+      bufferlist t;
+      t.substr_of(buf.bl, flush_offset - buf.bl_off, sizeof(File::WALFlush::WALLength));
+      dout(30) << "length dump\n";
+      t.hexdump(*_dout);
+      *_dout << dendl;
+      flush_length = *((uint64_t*)t.c_str());
+      dout(20) << __func__ << " flush_length " << flush_length << dendl;
+
+      if (flush_length == 0) {
+        dout(20) << __func__ << " flush end reached " << dendl;
+        break;
+      }
+      uint64_t increase = flush_length+(sizeof(File::WALFlush::WALLength)+sizeof(File::WALFlush::WALMarker));
+      file->fnode.size += increase;
+      file->wal_flushes.push_back({flush_offset, flush_length});
+      flush_offset += increase;
+    }
+  }
+}
+
+int64_t BlueFS::_read_wal(
```

Comment:
a style/pref comment: this is a pretty long function. Can it be broken down into logical, named, sub-functions?

Reply:
I might abstract some parts that are reused in different places but imho I prefer reading from top to bottom what this function does instead of going to defitinition in functions.